query-service-impl/src/main/java/org/hypertrace/core/query/service/postgres/PostgresClientFactory.java

```diff
@@ -58,32 +81,56 @@ public PostgresClient getPostgresClient(String clusterName) {
-      String url =
-          String.format(
-              "%s?user=%s&password=%s", clientConfig.getConnectionString(), user, password);
-      LOG.debug(
-          "Trying to create a Postgres client connected to postgres server using url: {}", url);
-      this.connection = DriverManager.getConnection(url);
+      String url = clientConfig.getConnectionString();
+      this.dataSource = createPooledDataSource(url, user, password);
     }
 
-    private PostgresClient(Connection connection) {
-      this.connection = connection;
+    private DataSource createPooledDataSource(String url, String user, String password) {
+      LOG.debug(
+          "Trying to create a Postgres client connected to postgres server using url: {}, user: {}, password: {}",
+          url,
+          user,
+          password);
+      ConnectionFactory connectionFactory = new DriverManagerConnectionFactory(url, user, password);
+      PoolableConnectionFactory poolableConnectionFactory =
+          new PoolableConnectionFactory(connectionFactory, null);
+      GenericObjectPool<PoolableConnection> connectionPool =
+          new GenericObjectPool<>(poolableConnectionFactory);
+      connectionPool.setMaxTotal(50);
+      connectionPool.setMaxIdle(10);
+      connectionPool.setMinIdle(5);
+      connectionPool.setMaxWaitMillis(5000);
+      poolableConnectionFactory.setPool(connectionPool);
+      poolableConnectionFactory.setValidationQuery("SELECT 1");
+      poolableConnectionFactory.setValidationQueryTimeout(5);
+      poolableConnectionFactory.setDefaultReadOnly(false);
+      poolableConnectionFactory.setDefaultAutoCommit(false);
+      poolableConnectionFactory.setDefaultTransactionIsolation(
+          Connection.TRANSACTION_READ_COMMITTED);
+      poolableConnectionFactory.setPoolStatements(false);
+      // poolableConnectionFactory.setMaxOpenPreparedStatements(100);
+      return new PoolingDataSource<>(connectionPool);
     }
 
-    public ResultSet executeQuery(String statement, Params params) throws SQLException {
-      PreparedStatement preparedStatement = buildPreparedStatement(statement, params);
-      return preparedStatement.executeQuery();
+    private PostgresClient(DataSource dataSource) {
+      this.dataSource = dataSource;
     }
 
-    private PreparedStatement buildPreparedStatement(String statement, Params params)
-        throws SQLException {
+    public Observable<Row> executeQuery(String statement, Params params) throws SQLException {
       String resolvedStatement = resolveStatement(statement, params);
-      return connection.prepareStatement(resolvedStatement);
+      try (Connection connection = dataSource.getConnection();
+          PreparedStatement preparedStatement = connection.prepareStatement(resolvedStatement);
+          ResultSet resultSet = preparedStatement.executeQuery()) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Query results: [ {} ]", resultSet);
+        }
+        return convert(resultSet);
```

Comment:
Shall we keep this as is? Result conversion as separate. As part of this class, let's have client related.

Reply:
We need to do conversion before result set is closed. We can move convert method to a separate class.